File: autonomy/camera-radar-fusion/src/shared/fusion/utils/calculate_distance_to_bounding_box.py
import numpy as np
from shared.camera.types.camera_detection import CameraDetection
from shared.fusion.utils.is_radar_inside_bounding_box import (
    is_radar_inside_bounding_box,
)
from shared.radar.types.radar_detection import RadarDetection


def calculate_distance_to_bounding_box(
    radar_detection: RadarDetection,
    camera_detection: CameraDetection,
) -> float:
    """
    Calculate minimum distance from radar point to bounding box edge.

    Args:
        radar_detection: Radar detection with image coordinates
        camera_detection: Camera detection with bounding box

    Returns:
        Minimum distance in pixels (0 if inside box)

    """
    if is_radar_inside_bounding_box(radar_detection, camera_detection):
        return 0.0

    x1, y1, x2, y2 = camera_detection.bounding_box
    radar_x = radar_detection.image_coordinate_x
    radar_y = radar_detection.image_coordinate_y

    # Calculate closest point on bounding box to radar point
    closest_x = np.clip(radar_x, x1, x2)
    closest_y = np.clip(radar_y, y1, y2)

    # Euclidean distance to closest point
    distance = np.sqrt((radar_x - closest_x) ** 2 + (radar_y - closest_y) ** 2)
    return float(distance)

File: autonomy/camera-radar-lidar-fusion/src/shared/fusion/utils/calculate_distance_to_bounding_box.rs
use crate::shared::camera::types::camera_detection::CameraDetection;
use crate::shared::fusion::utils::is_radar_inside_bounding_box::is_radar_inside_bounding_box;
use crate::shared::radar::types::radar_detection::RadarDetection;

pub fn calculate_distance_to_bounding_box(
    radar_detection: &RadarDetection,
    camera_detection: &CameraDetection,
) -> f64 {
    if is_radar_inside_bounding_box(radar_detection, camera_detection) {
        return 0.0;
    }

    let bounding_box = &camera_detection.bounding_box;
    let x1 = bounding_box[0];
    let y1 = bounding_box[1];
    let x2 = bounding_box[2];
    let y2 = bounding_box[3];

    let radar_x = radar_detection.image_coordinate_x;
    let radar_y = radar_detection.image_coordinate_y;

    let closest_x = radar_x.clamp(x1, x2);
    let closest_y = radar_y.clamp(y1, y2);

    ((radar_x - closest_x).powi(2) + (radar_y - closest_y).powi(2)).sqrt()
}
